internal/streamingnode/server/wal/interceptors/timetick/timetick_interceptor.go

```diff
@@ -0,0 +1,158 @@
+package timetick
+
+import (
+	"context"
+	"time"
+
+	"github.com/cockroachdb/errors"
+	"go.uber.org/zap"
+
+	"github.com/milvus-io/milvus/internal/streamingnode/server/wal/interceptors"
+	"github.com/milvus-io/milvus/internal/streamingnode/server/wal/interceptors/timetick/ack"
+	"github.com/milvus-io/milvus/pkg/log"
+	"github.com/milvus-io/milvus/pkg/streaming/util/message"
+)
+
+var _ interceptors.AppendInterceptor = (*timeTickAppendInterceptor)(nil)
+
+// timeTickAppendInterceptor is a append interceptor.
+type timeTickAppendInterceptor struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	ready  chan struct{}
+
+	ackManager *ack.AckManager
+	ackDetails *ackDetails
+	sourceID   int64
+}
+
+// Ready implements AppendInterceptor.
+func (impl *timeTickAppendInterceptor) Ready() <-chan struct{} {
+	return impl.ready
+}
+
+// Do implements AppendInterceptor.
+func (impl *timeTickAppendInterceptor) DoAppend(ctx context.Context, msg message.MutableMessage, append interceptors.Append) (msgID message.MessageID, err error) {
+	if msg.MessageType() != message.MessageTypeTimeTick {
+		// Allocate new acker for message.
+		var acker *ack.Acker
+		if acker, err = impl.ackManager.Allocate(ctx); err != nil {
+			return nil, errors.Wrap(err, "allocate timestamp failed")
+		}
+		defer func() {
+			acker.Ack(ack.OptError(err))
+			impl.ackManager.AdvanceLastConfirmedMessageID(msgID)
+		}()
+
+		// Assign timestamp to message and call append method.
+		msg = msg.
+			WithTimeTick(acker.Timestamp()).                  // message assigned with these timetick.
+			WithLastConfirmed(acker.LastConfirmedMessageID()) // start consuming from these message id, the message which timetick greater than current timetick will never be lost.
+	}
+	return append(ctx, msg)
+}
+
+// Close implements AppendInterceptor.
+func (impl *timeTickAppendInterceptor) Close() {
+	impl.cancel()
+}
+
+// execute start a background task.
+func (impl *timeTickAppendInterceptor) executeSyncTimeTick(interval time.Duration, param interceptors.InterceptorBuildParam) {
+	underlyingWALImpls := param.WALImpls
+
+	logger := log.With(zap.Any("channel", underlyingWALImpls.Channel()))
+	logger.Info("start to sync time tick...")
+	defer logger.Info("sync time tick stopped")
+
+	// Send first timetick message to wal before interceptor is ready.
```

Comment:
Move the given code snippet into a separate method named tryToSyncTimeTickUntilReady.